File: solutions/3090. Maximum Length Substring With Two Occurrences/3090.cpp
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
 public:
  int maximumLengthSubstring(string_view s) {
    int maxLen = 0;
    int count[26] = {0};

    for (int l = 0, r = 0; r < s.length(); ++r) {
      ++count[s[r] - 'a'];

      while (count[s[r] - 'a'] > 2) {
        --count[s[l] - 'a'];
        ++l;
      }

      maxLen = max(maxLen, r - l + 1);
    }

    return maxLen;
  }
};

File: solutions/3090. Maximum Length Substring With Two Occurrences/3090.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
  def maximumLengthSubstring(self, s:str) -> int:
    maxLen = 0
    count = [0] * 26

    l = r = 0

    while r < len(s):
      count[ord(s[r]) - ord('a')] += 1

      while count[ord(s[r]) - ord('a')] > 2:
        count[ord(s[l]) - ord('a')] -= 1
        l += 1

      maxLen = max(maxLen, r - l + 1)
      r += 1

    return maxLen    